[smt] guard tuple API usage by is_tuple_ast_type()

fbrausse opened the pull request:
Fixes #691. The tuple API was incorrectly used for unions in `convert_terminal()` for symbols.

fbrausse committed:
[smt] fix #691: convert_terminal: guard tuple API usage by is_tuple_ast_type()

lucasccordeiro commented:
Thanks for fixing this issue, @fbrausse. 

Can I ask you to add a regression test and rebase this branch?

fbrausse committed:
[smt] fix boolector and bitwuzla handling of array of unions in cex

The previous assumption that arrays only contain BVs that can be handled
by get_by_value() is no longer true, as we have moved to encode unions
the same way, but those can only be decoded by get_by_ast().

fbrausse committed:
[regression] add test case from #691

fbrausse commented:
> Thanks for fixing this issue, @fbrausse.
> 
> Can I ask you to add a regression test and rebase this branch?

Indeed, I should've thought of that earlier. It did catch problems in the handling of arrays of unions for both, Bitwuzla and Boolector. Should be fixed by now, let's wait for CI.

lucasccordeiro commented:
@fbrausse: could you please rebase this branch?
